### DISCUSSION TITLE
Examples for OpenID Connect Protocol Mapper?

### DISCUSSION THREAD

**Discussion by akloster** [keycloak/keycloak#22678] [Q&A] [answered]
I'd like to explore the possibility to map stuff into access token claims using Javascript. I got a script policy set up, barely. But I'm unsure how the mapping would work. If I can't find any examples I guess I'd have to parse the Java code to see what is exposed.

Also: The only way to use Javascript providers I see right now is through deploying a jar, right? There is conflicting information on whether it is still possible to upload scripts in the admin console. The problem I have with this is that it takes quite a long time (for a script) to reload a change: I have to make the jar and restart keycloak.

**Comment by justin-tay** [comment, marked as answer]
I think since Keycloak 18 it's no longer possible to upload scripts in the admin console, I think for security reasons.

That said it can be tedious to develop a script in this way. For development testing, on your development Keycloak instance, you can enable the `oidc-script-based-protocol-mapper`.

- The `scripts` feature needs to be enabled
- Create a `oidc-script-based-protocol-mapper.jar` with the following content in `META-INF/services/org.keycloak.protocol.ProtocolMapper`
  ```
  org.keycloak.protocol.oidc.mappers.ScriptBasedOIDCProtocolMapper
  ```
- Deploy to `providers`
- When you login to your `master` realm under `Provider Info` under `protocol-mapper` you should see `oidc-script-based-protocol-mapper` as one of them

You should now be able to `Add mapper > By configuration > Script Mapper`.

![script-mapper](https://github.com/keycloak/keycloak/assets/49700559/ace04bcc-7ee8-40ac-97e8-0338125a7914)

For the script, the return result that you set on `exports` will be the claim value used. The claim name is configured on the mapper as `Token Claim Name`, this is the same if you deployed a script.

The following is a sample script that sets the claim values to the client role mappings of that client.
```javascript
var ArrayList = Java.type("java.util.ArrayList");
var claimValues = new ArrayList();
var client = keycloakSession.getContext().getClient();
var forEach = Array.prototype.forEach;
forEach.call(user.getClientRoleMappingsStream(client).toArray(), function(roleModel) {
  claimValues.add(roleModel.getName());
});
exports = claimValues;
```
* Multivalued: On

You can set it to a JSON String
```javascript
exports = "{ \"hello\": \"world\" }";
```
* Claim JSON Type: JSON

That said since you have access to the `token` variable you can also just manipulate it

```javascript
token.setAcr("claimValue");
token.getOtherClaims().put("claimName", "claimValue");
```

You can't seem to edit the script by clicking `Save` right after the first creation for some reason (it says `Could not create mapping: 'Protocol mapper exists with same name'`), but if you back out and come back in you seem to be able to edit the mapper.

Also when evaluating the outputs in `Clients > Client details > Client scopes > Evaluate`, I notice that you need to change the `Users` to see the token result when you update the mappers.

**Reply by lfvjimisola** [reply]
Does this still work with 26.4.x?

Even after reading documentation and issues, it seems as if both Java and JavaScript are (still) supported.
Are there different use cases for them? Is one preferable or even deprecated?

Is the JavaScript mapper still preview and the Java custom protocol mapper official?

JavaScript:

[ScriptBasedOIDCProtocolMapper](https://www.keycloak.org/docs-api/latest/javadocs/org/keycloak/protocol/oidc/mappers/ScriptBasedOIDCProtocolMapper.html)

[OpenID Connect Protocol Mapper](https://www.keycloak.org/docs/latest/server_development/index.html#openid-connect-protocol-mapper)

Java:

[Custom Protocol Mapper (Baeldung)](https://www.baeldung.com/keycloak-custom-protocol-mapper)

**Reply by lfvjimisola** [reply]
Here is a 
[oidc-script-based-protocol-mapper.jar.zip](https://github.com/user-attachments/files/22848285/oidc-script-based-protocol-mapper.jar.zip) (just remove the zip extension) for ScriptBasedOIDCProtocalMapper. It has been tested with keycloak 26.4.0.

**Comment by akloster** [comment]
Thank you! I'll have to try that. I used the way from the documentation, and ultimately figured it out. There the script mapper just needs to have an expression on the last line, for example returning a string. But you can't edit it without restarting keycloak.

I figured out how to run an http request in the mapper/policy, and a separate process for the handler may make things easier. Also for development it may be useful to use "load" to load a script from a file which could even change during runtime.
